website/src/pages/admin/edit/[id].tsx

```diff
@@ -0,0 +1,176 @@
+import {
+  Box,
+  Button,
+  Card,
+  CardBody,
+  Flex,
+  Stack,
+  Tab,
+  TabList,
+  TabPanel,
+  TabPanels,
+  Tabs,
+  Text,
+  useColorModeValue,
+} from "@chakra-ui/react";
+import { GetServerSideProps, InferGetServerSidePropsType } from "next";
+import Head from "next/head";
+import { useTranslation } from "next-i18next";
+import { serverSideTranslations } from "next-i18next/serverSideTranslations";
+import { lazy, Suspense, useEffect, useMemo, useState } from "react";
+import { AdminArea } from "src/components/AdminArea";
+import { AdminLayout } from "src/components/Layout";
+import { MessageLoading } from "src/components/Loading/MessageLoading";
+import { MessageTree } from "src/components/Messages/MessageTree";
+import { TrackedTextarea } from "src/components/Survey/TrackedTextarea";
+import { TwoColumnsWithCards } from "src/components/Survey/TwoColumnsWithCards";
+import { get, put } from "src/lib/api";
+import { API_ROUTES, ROUTES } from "src/lib/routes";
+import { Message, MessageWithChildren } from "src/types/Conversation";
+import useSWRImmutable from "swr/immutable";
+import { useBoolean } from "@chakra-ui/react";
+import { useRouter } from "next/router";
+
+const RenderedMarkdown = lazy(() => import("../../../components/Messages/RenderedMarkdown"));
+
+const EditMessage = ({ id }: InferGetServerSidePropsType<typeof getServerSideProps>) => {
+  const { t } = useTranslation(["common", "tasks"]);
+  const router = useRouter();
+
+  const [messageText, setMessageText] = useState("");
+  const [submitting, setSubmitting] = useBoolean(false);
+  const { data, isLoading, error } = useSWRImmutable<{ tree: MessageWithChildren | null; message?: Message }>(
+    `/api/admin/messages/${id}/tree`,
+    get,
+    { keepPreviousData: true }
+  );
+
+  useEffect(() => {
+    if (data && data.message) {
+      setMessageText(data.message.text);
+    }
+  }, [data]);
+
+  const previewContent = useMemo(
+    () => (
+      <Suspense fallback={messageText}>
+        <RenderedMarkdown markdown={messageText}></RenderedMarkdown>
+      </Suspense>
+    ),
+    [messageText]
+  );
+
+  const getThreadFromTree = (tree: MessageWithChildren | null, targetMessageId: string) => {
+    if (!tree) {
+      return null;
+    }
+
+    if (tree.id === targetMessageId) {
+      return tree;
+    }
+
+    const filtered_children: (MessageWithChildren | null)[] = tree.children
+      .map((child) => getThreadFromTree(child, targetMessageId))
+      .filter((child) => child);
+    return filtered_children.length ? { ...tree, children: filtered_children } : null;
+  };
+
+  const handleSubmit = async () => {
+    setSubmitting.on();
+    await put(API_ROUTES.ADMIN_EDIT_MESSAGE(id), { arg: messageText });
```

Comment:
would it be an option to use swr mutations?

https://swr.vercel.app/docs/mutation#useswrmutation

This would also eliminate the need for `isSubmitting`